[mlir] support dialect attribute translation to LLVM IR

ftynse opened the pull request:
Extend the `amendOperation` mechanism for translating dialect attributes attached to operations from another dialect when translating MLIR to LLVM IR. Previously, this mechanism would have no knowledge of the LLVM IR instructions created for the given operation, making it impossible for it to perform local modifications such as attaching operation-level metadata. Collect instructions inserted by the LLVM IR builder and pass them to `amendOperation`.

gysit commented on mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:
```diff
@@ -52,6 +52,7 @@
 #include "llvm/Transforms/Utils/Cloning.h"
 #include "llvm/Transforms/Utils/ModuleUtils.h"
 #include <optional>
+#include <type_traits>
```
nit: is this needed or is it a clangd artifact?

gysit commented on mlir/include/mlir/Target/LLVMIR/LLVMTranslationInterface.h:
```diff
@@ -40,19 +42,21 @@ class LLVMTranslationDialectInterface
-  /// translated. The hook is expected to use moduleTranslation to look up the
-  /// translation results and amend the corresponding IR constructs. Does
-  /// nothing and succeeds by default.
+  /// translated and accepts as second argument the list of LLVM IR instructions
+  /// that were constructed when translating the operation. The hook is expected
+  /// to use moduleTranslation to look up the translation results and amend the
+  /// corresponding IR constructs. Does nothing and succeeds by default.
```
```suggestion
  /// translated and accepts as second argument the list of LLVM IR instructions
  /// that were constructed when translating the operation. The hook is expected
  /// to use moduleTranslation to look up the translation results, primarily for
  /// functions and globals that are not LLVM IR instructions, and amend the
  /// corresponding IR constructs. Does nothing and succeeds by default.
```
nit: trying to clarify that moduleTranslation lookups are mainly used for functions and globals since they are not instructions (at least that was may undestanding?)

gysit commented on mlir/include/mlir/Target/LLVMIR/ModuleTranslation.h:
```diff
@@ -50,6 +51,108 @@ class DINodeAttr;
+    bool wasEnabled;
+  };
+
+  void enable(bool enabled) { this->enabled = enabled; }
```
```suggestion
  void setEnabled(bool enabled) { this->enabled = enabled; }
```
ultra nit:

gysit reviewed (approved):
Very nice, now the import is jealous again :) 

LGTM modulo minor comments that may apply or not...

ftynse committed:
[mlir] support dialect attribute translation to LLVM IR

Extend the `amendOperation` mechanism for translating dialect attributes
attached to operations from another dialect when translating MLIR to
LLVM IR. Previously, this mechanism would have no knowledge of the LLVM
IR instructions created for the given operation, making it impossible
for it to perform local modifications such as attaching operation-level
metadata. Collect instructions inserted by the LLVM IR builder and pass
them to `amendOperation`.

Note that this currently doesn't extend to OpenMP operations as they go
through OpenMPIRBuilder as it lacks the required extensibility hooks.
Specifically, it owns an instance if IRBuilder with default template
arguments making it impossible to use the custom inserter class needed
for interceting LLVM IR instructions after they are created.

ftynse commented:
@gysit PTAL. I went with a simpler approach as OpenMP construction is weird and we can't fully handle it anyway.

gysit commented on mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:
```diff
@@ -641,11 +748,15 @@ ModuleTranslation::convertOperation(Operation &op,
 
+  InstructionCapturingInserter::CollectionScope scope(
+      builder, nestedConvertOperationUsesCapturingBuilder);
+  CapturingBuilderFlagScope flagScope(*this, false);
```
Would it be possible to pass the isCapturingBuilder flag together with the builder into convertOperation/convertBlock? It seems the flag is only true if the convertOperation is directly called from convertOneFunction? I assume the problem is that these are public functions and adding private impl functions makes things also quite complex.

Overall the flag scope solution is fine. I was just wondering if there is a more explicit approach.

gysit reviewed (approved):
Still LGTM

It's a shame OpenMP is incompatible with the original approach. But it seems to ok to support the other use cases initially.

ftynse commented:
> It's a shame OpenMP is incompatible with the original approach. But it seems to ok to support the other use cases initially.

I think it OpenMPIRBuilder needs to be updated, I'll try sending a patch there.
